core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java

```diff
@@ -0,0 +1,255 @@
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            int rows = 0;
+            ReferenceGenomeGene existingGene = null;
+
+            existingGene = getGene(gene.getEntrezGeneId(), gene.getReferenceGenomeId());
+
+            if (existingGene == null) {
+                //add gene, referring to this genetic entity
+                con = JdbcUtil.getDbConnection(DaoGene.class);
+                pstmt = con.prepareStatement
+                    ("INSERT INTO `reference_genome_gene` (`ENTREZ_GENE_ID`, `REFERENCE_GENOME_ID`,`CHR`,`CYTOBAND`,`EXONIC_LENGTH`,`START`,`END`) "
+                        + "VALUES (?,?,?,?,?,?,?)");
+                pstmt.setLong(1, gene.getEntrezGeneId());
+                pstmt.setInt(2, gene.getReferenceGenomeId());
+                pstmt.setString(3, gene.getChr());
+                pstmt.setString(4, gene.getCytoband());
+                pstmt.setInt(5, gene.getExonicLength());
+                pstmt.setLong(6, gene.getStart());
+                pstmt.setLong(7, gene.getEnd());
+                rows += pstmt.executeUpdate();
+            }
+            else {
+                rows += updateGene(existingGene);
+            }
+            return rows;
+        } catch (SQLException e) {
+            throw new DaoException(e);
+        } finally {
+            JdbcUtil.closeAll(DaoGene.class, con, pstmt, rs);
+        }
+    }
+
+
+    /**
+     * Gets the Gene with the Specified Entrez Gene ID.
+     * For faster access, consider using DaoGeneOptimized.
+     *
+     * @param entrezGeneId ENTRZ GENE ID.
+     * @return Canonical Gene Object.
+     * @throws DaoException Database Error.
+     */
+    public ReferenceGenomeGene getGene(long entrezGeneId, int referenceGenomeId) throws DaoException {
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            con = JdbcUtil.getDbConnection(DaoReferenceGenomeGene.class);
+            pstmt = con.prepareStatement
+                ("SELECT * FROM `reference_genome_gene` WHERE `ENTREZ_GENE_ID` = ? AND `REFERENCE_GENOME_ID` = ?");
+            pstmt.setLong(1, entrezGeneId);
+            pstmt.setInt(2, referenceGenomeId);
+            rs = pstmt.executeQuery();
+            if (rs.next()) {
+                return extractGene(rs);
+            } else {
+                return null;
+            }
+        } catch (SQLException e) {
+            throw new DaoException(e);
+        } finally {
+            JdbcUtil.closeAll(DaoGene.class, con, pstmt, rs);
+        }
+    }
+
+    /**
+     * Gets the Gene with the Specified Entrez Gene ID.
+     * For faster access, consider using DaoGeneOptimized.
+     *
+     * @param hugoGeneSymbol Hugo Gene Symbol.
+     * @return Canonical Gene Object.
+     * @throws DaoException Database Error.
+     */
+    public ReferenceGenomeGene getGene(String hugoGeneSymbol, int referenceGenomeId) throws DaoException {
```

Comment:
On my first readthrough of this code, this join logic means that this function will be blind to any records in reference_genome_gene that do not have a matching (by entrez_gene_id) record in the table gene. So I am thinking that it always will be the case that when records are added to reference_genome_gene, there will be a matching record in table gene to link to. (maybe I am just stating the obvious here)

I guess I might recommend that in the code which adds records to reference_genome_gene, we might want to add a check that a matching record exists in `gene`. (or maybe we already have a foreign key constraint between tables ... I have not gotten there yet)

Reply:
yes, we have a foreign key constraint in the reference genome table links to the gene table, you will not be able to add to the reference gene table if no record found in the gene table, the beauty of foreign key :) .